File: data-structure-and-algorithm/jian-zhi-offer/0068-2.Lowest-Common-Ancestor/LowestCommonAncestor2.java
public class LowestCommonAncestor2 {
    public static TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null || p == root || q == root){
            return root;
        }
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if(left == null){
            return right;
        }
        if(right == null){
            return left;
        }
        return root;
    }

	private static class TreeNode {
		int val;
		LowestCommonAncestor2.TreeNode left;
		LowestCommonAncestor2.TreeNode right;
		TreeNode(int x) { val = x; }
	}

    public static void main(String[] args) {

    }
}

File: data-structure-and-algorithm/jian-zhi-offer/0068-2.Lowest-Common-Ancestor/LowestCommonAncestor2.py
# Definition for a binary tree node.
class TreeNodeTraversal:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class LowestCommonAncestor2:
    def lowestCommonAncestor(self, root: TreeNodeTraversal, p: TreeNodeTraversal, q: TreeNodeTraversal) -> TreeNodeTraversal:
        if root is None or p == root or q == root:
            return root
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        if left is None:
            return right
        if right is None:
            return left
        return root


if __name__ == '__main__':
    LowestCommonAncestor2()
